MSAL/test/integration/native_auth/end_to_end/otp_code_retriever/MSALNativeAuthEmailCodeRetriever.swift

```diff
@@ -0,0 +1,115 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.  
+
+import XCTest
+
+class MSALNativeAuthEmailCodeRetriever: XCTestCase {
+
+    private let baseURLString = "https://www.1secmail.com/api/v1/?action="
+    private let secondsToWait = 4.0
+    private let numberOfRetry = 3
+
+    func retrieveEmailOTPCode(email: String) async -> String? {
+        let comps = email.components(separatedBy: "@")
+        guard comps.count == 2 else {
+            XCTFail("Invalid email address")
+            return nil
+        }
+        let local = comps[0]
+        let domain = comps[1]
+
+        guard let lastMessageId = await retrieveLastMessage(local: local, domain: domain, retryCounter: numberOfRetry) else {
+            XCTFail("Something went wrong retrieving the messages for the email specified")
+            return nil
+        }
+        guard let emailOTPCode = await retrieveOTPCodeFromMessage(local: local, domain: domain, messageId: lastMessageId) else {
+            XCTFail("Something went wrong retrieving the OTP code")
+            return nil
+        }
+        return emailOTPCode
+    }
+    
+    func generateRandomEmailAddress() -> String {
+        let randomId = UUID().uuidString.prefix(8)
+        return "native-auth-signup-\(randomId)@1secmail.org"
+    }
+
+    private func retrieveLastMessage(local: String, domain: String, retryCounter: Int) async -> Int? {
+        guard retryCounter > 0, let url = URL(string: baseURLString + "getMessages&login=\(local)&domain=\(domain)") else {
+            return nil
+        }
+        try? await Task.sleep(nanoseconds: UInt64(secondsToWait * Double(NSEC_PER_SEC)))
+        var request = URLRequest(url: url)
+        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
+        do {
+            let (data, response) = try await URLSession.shared.data(for: request)
+            
+            let code = (response as? HTTPURLResponse)?.statusCode ?? 0
+            guard let httpResponse = response as? HTTPURLResponse,
+                    httpResponse.statusCode == 200 else {
+                print("Unexpected response from 1secmail: \(code) status code")
+                return nil
+            }
+            guard var dataDictionary = try JSONSerialization.jsonObject(with: data, options: []) as? [[String: Any]] else {
+                return nil
+            }
+            if dataDictionary.count > 0 {
+                dataDictionary.sort(by: {($0["id"] as? Int ?? 0) > ($1["id"] as? Int ?? 0)})
+                return dataDictionary.first?["id"] as? Int
+            } else {
+                // log only for the final retry
+                if (retryCounter == 1) {
```

Comment:
Can you do this check comparing it to the `numberOfRetry` variable? Just in case we decide to update the number of retries in the future, to avoid changing it here as well.

Reply:
Oh right, sorry my bad! 🤦‍♂️